File: src/test/java/kitchenpos/application/OrderServiceCompleteTest.java
package kitchenpos.application;

import static kitchenpos.application.constant.KitchenposTestConst.TEST_ORDER_TABLE_NUMBER_OF_GUEST;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.NoSuchElementException;
import java.util.UUID;
import kitchenpos.application.fakerepository.MenuFakeRepository;
import kitchenpos.application.fakerepository.OrderFakeRepository;
import kitchenpos.application.fakerepository.OrderTableFakeRepository;
import kitchenpos.domain.MenuRepository;
import kitchenpos.domain.Order;
import kitchenpos.domain.OrderRepository;
import kitchenpos.domain.OrderStatus;
import kitchenpos.domain.OrderTable;
import kitchenpos.domain.OrderTableRepository;
import kitchenpos.domain.OrderType;
import kitchenpos.infra.KitchenridersClient;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
@DisplayName("OrderService 클래스의 complete메소드 테스트")
@ExtendWith(MockitoExtension.class)
public class OrderServiceCompleteTest {

    private OrderRepository orderRepository;
    private MenuRepository menuRepository;
    private OrderTableRepository orderTableRepository;

    @Mock
    private KitchenridersClient mockRidersClient;

    private OrderService sut;

    @BeforeEach
    void setUp() {
        orderRepository = new OrderFakeRepository();
        menuRepository = new MenuFakeRepository();
        orderTableRepository = new OrderTableFakeRepository();

        sut = new OrderService(orderRepository, menuRepository,
            orderTableRepository, mockRidersClient);
    }

    @Test
    void 주문이_없으면_예외를_발생시킨다() {

        // when & then
        assertThatThrownBy(() -> sut.completeDelivery(UUID.randomUUID()))
            .isExactlyInstanceOf(NoSuchElementException.class);
    }


    @Test
    void 배달주문일_때_주문의_상태가_delivered가_아니면_예외를_발생시킨다() {
        // given
        final Order deliveringOrder = orderRepository.save(
            createDelieveryOrder(OrderStatus.DELIVERING));
        final Order completedOrder = orderRepository.save(
            createDelieveryOrder(OrderStatus.COMPLETED));

        // when & then
        assertThatThrownBy(() -> sut.complete(deliveringOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);

        assertThatThrownBy(() -> sut.complete(completedOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);
    }


    @Test
    void 매장주문일_때_served가_아니면_예외를_발생시킨다() {
        // given
        final Order acceptedOrder = orderRepository.save(createEatinOrder(OrderStatus.ACCEPTED));
        final Order completedOrder = orderRepository.save(createEatinOrder(OrderStatus.COMPLETED));

        // when & then
        assertThatThrownBy(() -> sut.complete(acceptedOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);

        assertThatThrownBy(() -> sut.complete(completedOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);
    }


    @Test
    void 포장주문일_때_served가_아니면_예외를_발생시킨다() {
        // given
        final Order acceptedOrder = orderRepository.save(createTakeoutOrder(OrderStatus.ACCEPTED));
        final Order completedOrder
            = orderRepository.save(createTakeoutOrder(OrderStatus.COMPLETED));

        // when & then
        assertThatThrownBy(() -> sut.complete(acceptedOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);

        assertThatThrownBy(() -> sut.complete(completedOrder.getId()))
            .isExactlyInstanceOf(IllegalStateException.class);
    }

    @Test
    void 포장주문일_때_주문의_상태를_completed로_변경하여_반환한다() {
        // given
        final Order order = orderRepository.save(createTakeoutOrder(OrderStatus.SERVED));

        // when
        final Order actual = sut.complete(order.getId());

        // then
        assertThat(actual.getStatus()).isEqualTo(OrderStatus.COMPLETED);
    }


    @Test
    void 배달주문일_때_주문의_상태를_completed로_변경하여_반환한다() {
        // given
        final Order order = orderRepository.save(createDelieveryOrder(OrderStatus.DELIVERED));

        // when
        final Order actual = sut.complete(order.getId());

        // then
        assertThat(actual.getStatus()).isEqualTo(OrderStatus.COMPLETED);
    }


    @Test
    void 매장주문일_때_주문의_상태를_completed로_변경하여_반환한다() {
        // given
        final OrderTable orderTable = orderTableRepository.save(createOccupiedOrderTable());
        final Order order = orderRepository.save(createEatinOrderWithOrderTable(orderTable));

        // when
        final Order actual = sut.complete(order.getId());

        // then
        assertThat(actual.getStatus()).isEqualTo(OrderStatus.COMPLETED);
    }

    @Test
    void 매장주문일_때_주문에_할당된_매장테이블을_치운다() {
        // given
        final OrderTable orderTable = orderTableRepository.save(createOccupiedOrderTable());
        final Order order = orderRepository.save(createEatinOrderWithOrderTable(orderTable));

        // when
        sut.complete(order.getId());

        // then
        assertThat(orderTable.getNumberOfGuests()).isZero();
        assertThat(orderTable.isOccupied()).isFalse();
    }


    final OrderTable createOccupiedOrderTable() {
        final OrderTable orderTable = new OrderTable();
        orderTable.setNumberOfGuests(TEST_ORDER_TABLE_NUMBER_OF_GUEST);
        orderTable.setOccupied(true);

        return orderTable;
    }

    final Order createEatinOrderWithOrderTable(final OrderTable orderTable) {
        final Order order = new Order();

        order.setStatus(OrderStatus.SERVED);
        order.setType(OrderType.EAT_IN);
        order.setOrderTable(orderTable);

        return order;
    }

    final Order createDelieveryOrder(final OrderStatus status) {
        return create(OrderType.DELIVERY, status);
    }

    final Order createEatinOrder(final OrderStatus status) {
        return create(OrderType.EAT_IN, status);
    }

    final Order createTakeoutOrder(final OrderStatus status) {
        return create(OrderType.TAKEOUT, status);
    }

    final Order create(final OrderType type, final OrderStatus status) {
        final Order order = new Order();
        order.setType(type);
        order.setStatus(status);

        return order;
    }
}

File: src/test/java/kitchenpos/application/OrderServiceCompleteTest.kt
package kitchenpos.application

import io.kotest.assertions.throwables.shouldThrowAny
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeTypeOf
import kitchenpos.domain.OrderRepository
import kitchenpos.domain.OrderStatus.ACCEPTED
import kitchenpos.domain.OrderStatus.COMPLETED
import kitchenpos.domain.OrderStatus.DELIVERED
import kitchenpos.domain.OrderStatus.DELIVERING
import kitchenpos.domain.OrderStatus.SERVED
import kitchenpos.domain.OrderTableRepository
import kitchenpos.domain.OrderType.DELIVERY
import kitchenpos.domain.OrderType.EAT_IN
import kitchenpos.domain.OrderType.TAKEOUT
import kitchenpos.infra.KitchenridersClient
import kitchenpos.testsupport.FakeMenuRepository
import kitchenpos.testsupport.FakeOrderRepository
import kitchenpos.testsupport.FakeOrderTableRepository
import kitchenpos.testsupport.OrderFixtures
import kitchenpos.testsupport.OrderTableFixtures

class OrderServiceCompleteTest : ShouldSpec({
    lateinit var orderTableRepository: OrderTableRepository
    lateinit var orderRepository: OrderRepository
    lateinit var service: OrderService

    beforeTest {
        orderTableRepository = FakeOrderTableRepository()
        orderRepository = FakeOrderRepository()

        service = OrderService(
            orderRepository,
            FakeMenuRepository(),
            orderTableRepository,
            KitchenridersClient()
        )
    }

    context("주문 완료") {
        context("배달 주문") {
            should("성공") {
                // given
                val deliveredOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = DELIVERY,
                        status = DELIVERED,
                        menus = listOf()
                    )
                )

                // when
                val order = service.complete(deliveredOrder.id)

                // then
                order.status shouldBe COMPLETED
            }

            should("실패 - 주문이 배달 완료 상태가 아닌 경우") {
                // given
                val deliveredOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = DELIVERY,
                        status = DELIVERING,
                        menus = listOf()
                    )
                )

                // when
                val exception = shouldThrowAny {
                    service.complete(deliveredOrder.id)
                }

                // then
                exception.shouldBeTypeOf<IllegalStateException>()
            }
        }

        context("포장 주문") {
            should("성공") {
                // given
                val servedOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = TAKEOUT,
                        status = SERVED,
                        menus = listOf()
                    )
                )

                // when
                val order = service.complete(servedOrder.id)

                // then
                order.status shouldBe COMPLETED
            }

            should("실패 - 주문 전달 상태가 아닌 경우") {
                // given
                val acceptedOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = TAKEOUT,
                        status = ACCEPTED,
                        menus = listOf()
                    )
                )

                // when
                val exception = shouldThrowAny {
                    service.complete(acceptedOrder.id)
                }

                // then
                exception.shouldBeTypeOf<IllegalStateException>()
            }
        }

        context("매장 주문") {
            should("성공") {
                // given
                val occupiedOrderTable = orderTableRepository.save(
                    OrderTableFixtures.createOrderTable()
                        .apply { isOccupied = true }
                )
                val acceptedOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = EAT_IN,
                        status = SERVED,
                        menus = listOf(),
                        orderTable = occupiedOrderTable
                    )
                )

                // when
                val order = service.complete(acceptedOrder.id)

                // then
                order.status shouldBe COMPLETED
            }

            should("실패 - 주문 전달 상태가 아닌 경우") {
                // given
                val acceptedOrder = orderRepository.save(
                    OrderFixtures.createOrder(
                        type = EAT_IN,
                        status = ACCEPTED,
                        menus = listOf(),
                        orderTable = orderTableRepository.save(
                            OrderTableFixtures.createOrderTable()
                        )
                    )
                )

                // when
                val exception = shouldThrowAny {
                    service.complete(acceptedOrder.id)
                }

                // then
                exception.shouldBeTypeOf<IllegalStateException>()
            }
        }
    }
})
